server/entitymanager/entitymanager.go

```diff
@@ -92,10 +146,70 @@ func (m *InMemoryEntityManager) SetStatus(req *bootz.ReportStatusRequest) error
 	return nil
 }
 
+// // readKeyPair reads the cert/key pair from the specified directory.
+// Certs must have the format {name}_pub.pem and keys must have the format {name}_priv.pem
+func readKeypair(dir, name string) (*service.KeyPair, error) {
+	cert, err := os.ReadFile(filepath.Join(dir, fmt.Sprintf("%v_pub.pem", name)))
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %v cert: %v", name, err)
+	}
+	key, err := os.ReadFile(filepath.Join(dir, fmt.Sprintf("%v_priv.pem", name)))
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %v key: %v", name, err)
+	}
+	return &service.KeyPair{
+		Cert: string(cert),
+		Key:  string(key),
+	}, nil
+}
+
+// generateServerTlsCert uses the PDC key as the server certificate.
+func loadServerTlsCert(pdc *service.KeyPair) (*tls.Certificate, error) {
+	tlsCert, err := tls.X509KeyPair([]byte(pdc.Cert), []byte(pdc.Key))
+	if err != nil {
+		return nil, fmt.Errorf("unable to load PDC keys %v", err)
+	}
+	return &tlsCert, err
+}
+
+// parseSecurityArtifacts reads from the specified directory to find the required keypairs and ownership vouchers.
+func parseSecurityArtifacts(artifactDir string) (*service.SecurityArtifacts, error) {
+	oc, err := readKeypair(artifactDir, "oc")
+	if err != nil {
+		return nil, err
+	}
+	pdc, err := readKeypair(artifactDir, "pdc")
+	if err != nil {
+		return nil, err
+	}
+	vendorCA, err := readKeypair(artifactDir, "vendorca")
+	if err != nil {
+		return nil, err
+	}
+	// use pdc key as server cer
+	tlsCert, err := loadServerTlsCert(pdc)
+	if err != nil {
+		return nil, err
+	}
+	return &service.SecurityArtifacts{
+		OC:         oc,
+		PDC:        pdc,
+		VendorCA:   vendorCA,
+		TLSKeypair: tlsCert,
+	}, nil
+}
+
 // Sign unmarshals the SignedResponse bytes then generates a signature from its Ownership Certificate private key.
-func (m *InMemoryEntityManager) Sign(resp *bootz.GetBootstrapDataResponse, serial string) error {
+func (m *InMemoryEntityManager) Sign(resp *bootz.GetBootstrapDataResponse, chassis *service.EntityLookup, controllerCard string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// get security artifacts for the device
+	secArtifact, err := parseSecurityArtifacts(m.defaults.ArtifactDir)
```

Comment:
I don't think we should be parsing these files each time Sign() is called. This is a potentially expensive process as it involves I/O. Could we not just parse artifacts once and store them?

Reply:
yes, I was thinking doing the same. will change the code.